sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                tmin = min([x[4] for x in h_ep])
-                for idx, ep in enumerate(h_ep):
-                    #print(idx, (ep[0], ep[1]), ep[2], ep[3][ep[0]:ep[1]+1])
-                    xax = list(range(ep[0], ep[1]+1))
-                    yax = [mcats.index(ep[2])]*len(xax)
-                    thickness  = ep[4]
-                    lsize = ((thickness - tmin)/ (tmax - tmin)) * (5-0.5) + 0.5 if (tmax - tmin) != 0.0 else 0.5
-                    #lsize = np.log(thickness) + 1 TODO: Either take log or normalize between [0.5 5]
-                    msize = (lsize*2)+1
-                    ax.plot(xax, yax, color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], linewidth=lsize)
-                    ax.plot(ep[0], mcats.index(ep[2]), color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], marker='.', linewidth=0, markersize=msize, label=micro2macro[micro[ep[2]]])
-                    ax.plot(ep[1], mcats.index(ep[2]), color=mcols[macro_inv[micro2macro[micro[ep[2]]]]], marker='.', linewidth=0, markersize=msize)
-                    plt.yticks(range(len(mcats)), mcats_lab, rotation='0')
-                legend_without_duplicate_labels(ax)
-                plt.grid(True, alpha=0.4)
-                
-                #plt.tight_layout()
-                #plt.savefig('Pres-Micro-attack-episodes-Team'+str(tid) +'-Connection'+ attacker[0]+'--'+attacker[1]+'.png')
-                plt.show()
-        
-        team_episodes.append(t_ep)
+            host_episodes.sort(key=lambda tup: tup[0])
+            team_host_episodes[attacker_victim] = host_episodes
+
+            if plot_alert_volumes:
+                _plot_alert_volume_per_episode(tid, attacker_victim, host_episodes, mcats)
+
+        _team_episodes.append(team_host_episodes)
         team_times.append(_team_times)
-    return (team_episodes, team_times)
-            
-## 17
+    return _team_episodes, team_times
 
-#### Host = [connections] instead of team level representation
-def host_episode_sequences(team_episodes):
-    host_data = {}
-    #team_host_data= []
-    print('# teams ', len(team_episodes))
+
+# Step 3: Create episode sequences
+# Host = [connections] instead of team level representation
+def host_episode_sequences(_team_episodes):
+    _host_data = {}
+    print('# teams:', len(_team_episodes))
     print('----- TEAMS -----')
-    for tid, team in enumerate(team_episodes):
-        print(tid, sep=' ', end=' ', flush=True )
-        #print(len(set([x[0] for x in team.keys()])))
-        for attacker,episodes in team.items():
-            #print(attacker)
+    for tid, team in enumerate(_team_episodes):
+        print(tid, sep=' ', end=' ', flush=True)
+        for (attacker, victim), episodes in team.items():
             if len(episodes) < 2:
                 continue
-            perp= attacker[0]
-            vic = attacker[1]
-            #print(perp)
-            #if ('10.0.0' in perp or '10.0.1' in perp):
+            # if ('10.0.0' in attacker or '10.0.1' in attacker):
```

Comment:
I guess, this is for one of the three datasets? Leaving like that?

Reply:
Yes, testing for specific cases